File: code/perfect-java-code/perfect-design-pattern/src/creation/factory/factorymethod/pizzastore/order/OrderPizza.java
package creation.factory.factorymethod.pizzastore.order;


import creation.factory.factorymethod.pizzastore.pizza.Pizza;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

/**
 * @author Kelly
 * @create 2020-09-20 14:51
 */
public abstract class OrderPizza {
    // 定义一个抽象方法 createPizza, 让各个工厂子类自己去实现
    abstract Pizza createPizza(String orderType);

    public OrderPizza() {
        Pizza pizza = null;
        String orderType;
        do {
            orderType = getType();
            pizza = createPizza(orderType);
            pizza.prepare();
            pizza.bake();
            pizza.cut();
            pizza.box();

        } while (true);
    }

    private String getType() {
        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            System.out.println("Input pizza type: ");
            String type = null;
            type = br.readLine();
            return type;
        } catch (IOException e) {
            e.printStackTrace();
            return "";
        }
    }

}

File: code/perfect-java-code/perfect-scala/src/main/scala/com/learn/chapter17/simplefactory/pizzastore/order/OrderPizza.scala
package com.learn.chapter17.simplefactory.pizzastore.order

import com.learn.chapter17.simplefactory.pizzastore.pizza.Pizza

import scala.io.StdIn
import scala.util.control.Breaks.{break, breakable}

/**
 * @author Kelly
 * @create 2021-05-13 15:42
 */
class OrderPizza {
  var orderType: String = _
  var pizza: Pizza = _

  breakable {
    do {
      println("请输入 pizza 的类型, 使用简单工厂模式~~")
      orderType = StdIn.readLine()
      pizza = SimpleFactory.createPizza(orderType)
      if (pizza == null) {
        break()
      }
      this.pizza.prepare()
      this.pizza.bake()
      this.pizza.cut()
      this.pizza.box()
    } while (true)
  }
}
